### DISCUSSION TITLE
Caching with custom attributes

### DISCUSSION THREAD

**Discussion by Naghal** [spatie/laravel-permission#2170] [Q&A] [answered]
I am starting this conversation as I am unsure if it this something that is supported or a bug.
I have added a column named "organisation_id" to my Role model which extends the base Role class.
The problem I intended to solve at first was to have a role with the same name in different `Organisation`.
As an example, 
- In organisation A, there is a role named "Admin" with permission 1,2,3.
- In organisation B, there is a role name "Admin" with permission 3,4.

The issue I am running into is it seems that the cache only store one role at a time because the name is the same.
So let's say I log in with user A from organisation a with role admin, everything is fine, but when I log in with user B from organisation b and role admin, he has no permission (Should have 3 and 4).
Also, I see a `organisation_id` key in the cache aliased to `l`, but it is always empty. 

**Role.php**
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Support\Facades\Auth;
use Spatie\Permission\Exceptions\RoleAlreadyExists;
use Spatie\Permission\Models\Role as BaseRole;
use Spatie\Permission\Guard;

class Role extends BaseRole
{
    protected static function booted()
    {
        static::addGlobalScope('organisation', function (Builder $builder) {
            if (Auth::user() && Auth::user()->su === 0) {
                $builder->where('organisation_id', Auth::user()->organisation_id);
            }
        });
    }

    protected $casts = [
        'updated_at' => 'datetime:Y-m-d H:i:s',
        'created_at' => 'datetime:Y-m-d H:i:s',
    ];

    public function organisation()
    {
        return $this->belongsTo(Organisation::class);
    }

    public static function create(array $attributes = [])
    {
        $attributes['guard_name'] = $attributes['guard_name'] ?? Guard::getDefaultName(static::class);

        if (static::where('name', $attributes['name'])
            ->where('guard_name', $attributes['guard_name'])
            ->where('organisation_id', $attributes['organisation_id'])
            ->first()) {
            throw RoleAlreadyExists::create($attributes['name'], $attributes['guard_name']);
        }

        return static::query()->create($attributes);
    }
}
```

![image](https://user-images.githubusercontent.com/29404633/188702723-241661a4-9f20-48d8-87c6-b0961075d454.png)

![Sélection_013](https://user-images.githubusercontent.com/29404633/188702629-edc4910c-490b-4626-9782-ce2cbe78e01d.png)

How can I solve the cache issue?
Using v 5.5.5

**Comment by Naghal** [comment]
@erikn69 I see you have made a fix for that kind of problem, can you spot anything wrong in my role class?

**Comment by erikn69** [comment, marked as answer]
First, the `organisation_id` seems like `team_id` feature
https://github.com/spatie/laravel-permission/blob/dbc6c18ef4b2d097350a5faf82382564748e6c97/config/permission.php#L96

>seems that the cache only store one role at a time because the name is the same

Second, the name can be the same if you use the teams feature; Otherwise it could be prone to failure.

But if you think it's an issue, make a demo app to check
https://spatie.be/docs/laravel-permission/v5/basic-usage/new-app#content-creating-a-demo-app

**Reply by Naghal** [reply]
You are right, I have seen the team feature (which was released after I had mades all the changes).
I have tried to use this solution, but I could not get the migration to run.
It always results in a `Segmentation fault (core dumped)`.
I have copied the latest permission.php file in my project and set the following key:
```php
'team_foreign_key' => 'organisation_id',
'teams' => true,
```
Then, I ran the commands `php artisan permission:setup-teams` and `php artisan migrate`.

**Edit**
After further debugging, I solved the issue by adding the proper primary key name to be dropped.
`$table->dropPrimary('model_has_permissions_permission_model_type_primary');`
`$table->dropPrimary('model_has_roles_role_model_type_primary');`

**Reply by erikn69** [reply]
Take a look to #1804, #1944 and other PR about Teams, maybe you will find the problem on your implementation

**Reply by Naghal** [reply]
From my testing so far, this seems like the correct solution.
All I have to find now is a way to be able to display roles of users from all organisations.
Thanks for the help!
